compiler/runtime.go

```diff
@@ -698,6 +722,7 @@ func (f *functionValue) Call(ctx context.Context, args values.Object) (values.Va
 			a = v
 		}
 		scope.Set(p.Key, a)
+		a.Retain()
```

Comment:
This retain may be unnecessary since we are evaluating the arguments and then immediately transferring ownership to the scope.

We do need to release the values we have included in the scope. I think we need to debug this as this may be another circumstance similar to the one above where retain is not needed. Contrary to the one above, I think release will be needed on the values we add to the scope because in this circumstance we're the owners rather than borrowing.

Reply:
I'm trying to understand the example of `(r) => ({foo: r.a, bar: r.a})` works, since the array `r.a` that is in the input is referenced twice in the output, so I would expect that array to have a reference count of 2 once the input has been released. This seems important to avoid releasing too many times down the line.

So how does that happen? I see that in `memberEvaluator` we extract the array from the record here:
https://github.com/influxdata/flux/blob/0ee7a903a4daa146f62cd4c4f3adc34a28c33a62/compiler/runtime.go#L574
But I don't see how that increments the reference count.

Looking at where we create the output object in `objEvaluator` I also don't see how the reference count would be incremented there. So I'm confused on this point.

Reply:
This is a good catch. Both array evaluator and object evaluator need to retain the reference it retrieves.

Reply:
Just so I'm clear on what's being requested, we want to:

- change the `Retain` call on line 724 to a `Release` call
- add `Retain` calls to the `Get` and `Set` methods of scopes, objects, and arrays

Do I have that right?

Reply:
I don't think you change it to a release. I may be able to walk through each of these with you tomorrow though.

The retains should be in the evaluators not the underlying interface implementation.